Enhancement/remove duplicate artifact 67

richpsharp opened the pull request:
Removed `copy_duplicate_artifact` and `hardlink_allowed` parameters and functionality from TaskGraph. This is to address a design error that TaskGraph is not well suited for caching file results to avoid recomputation.

Fixes #67

phargogh commented on taskgraph/Task.py:
```diff
@@ -1255,8 +1167,8 @@ def is_precalculated(self):
             self._store_result,
-            # the x[2] is to only take the *hash* part of the 'file_stat'
-            str([x[2] for x in file_stat_list]))
+            # the x[1] is to only take the *hash* part of the 'file_stat'
```
The "hash" part of this comment was a bit perplexing to me, since if we're using `"sizetimestamp"` as the `target_hash_algorithm`, this will end up being a string that isn't a digest at all.  I'm not sure if it's a bug in this comment specifically or a bug in the docstring of `_hash_file` (though that docstring specifically refers to a hash digest, which isn't strictly true), but what do you think?

richpsharp replied:
I'm okay with this since "sizetimestamp" is a hash too -- it changes one kind of data into a fingerprint. But also note it won't always be "sizetimestamp", it'll be whatever the value of `hash_algorithm` is. The point is `x[1]` is the hash component of the tuple that's returned from `_get_file_stats` and I otherwise thought it would be confusing if I left that `x[1]` hanging without comment.

phargogh commented on taskgraph/Task.py:
```diff
@@ -1407,11 +1319,10 @@ def _get_file_stats(
                     not ignore_directories) and os.path.exists(norm_path):
                 if hash_algorithm == 'exists':
-                    yield (norm_path, 'exists', 'exists')
+                    yield (norm_path, 'exists')
```
Could you update the docstring to reflect the updated tuple structure?

richpsharp replied:
Updated in 1d5314f

phargogh reviewed (changes requested):
Thanks @richpsharp !  I had a couple comments about comments, and then I also noticed that the `hash_algorithm` note in `Task.__init__()` doesn't mention `exists` as a valid option, though the source code obviously handles that case.  Should that be described in the docstring as well?  I know it's a little out of the scope of this PR, but it came up while I was looking through things.

richpsharp commented:
> Thanks @richpsharp ! I had a couple comments about comments, and then I also noticed that the `hash_algorithm` note in `Task.__init__()` doesn't mention `exists` as a valid option, though the source code obviously handles that case. Should that be described in the docstring as well? I know it's a little out of the scope of this PR, but it came up while I was looking through things.

Oh great, yes i added 'exists' a while back and it's documented some places but not all the way up. I added that to `add_task` and `__init__` in bd77185
